README.md
# video_meta_info

A flutter plugin can retrieve video meta info.


## Getting Started

This project is a starting point for a Flutter
[plug-in package](https://flutter.dev/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## 1.0.1
* Change the minSdkVersion to 28 to be compatible old app

## 1.0.0
* Update package name to avoid compile issues.

## 0.0.1
* Publish the package on pub.dev


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:video_meta_info_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/video_meta_info_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:video_meta_info/video_meta_info.dart';

void main() {
  const MethodChannel channel = MethodChannel('video_meta_info');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await VideoMetaInfo.platformVersion, '42');
  });
}


example/lib/main.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';
import 'package:file_selector/file_selector.dart';
import 'package:permission_handler/permission_handler.dart';

import 'package:video_meta_info/video_meta_info.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  final videoInfo = VideoMetaInfo();

  @override
  void initState() {
    if (Platform.isAndroid) {
      Permission.contacts.request();
    }
    super.initState();
  }

  String info = "";
  String videoFilePath = "";

  getVideoInfo() async {

    /// here file path of video required
    if (Platform.isIOS) {
      videoFilePath =
          "/Users/User/Library/Developer/CoreSimulator/Devices/6A0D4244-1DEB-49C3-9837-C08E19DAED31/data/Media/DCIM/100APPLE/IMG_0011.mp4";
    } else if (Platform.isAndroid) {
      videoFilePath = "storage/emulated/0/Geocam/Videos/4.mp4";
    }

    // var file;
    XFile? file;

    if (Platform.isMacOS) {
      final typeGroup = XTypeGroup(label: 'videos', extensions: ['mov', 'mp4']);
      file = await openFile(acceptedTypeGroups: [typeGroup]);
    } else {

      XFile? f = await ImagePicker().pickVideo(
          source: ImageSource.gallery, maxDuration: const Duration(seconds: 302));

      if (f!=null) {
        file = f;
      }

      // final FilePickerResult? result = await FilePicker.platform.pickFiles(
      //   type: FileType.video,
      // );

      // final PlatformFile? file = result!.files.first;


      // file = result!.files.first; // File(pickedFile!.path);
    }
    if (file == null) {
      return;
    }

    var a = await videoInfo.getVideoInfo(file.path);
    setState(() {
      info =
          "title=> ${a?.title}\npath=> ${a?.path}\nauthor=> ${a?.author}\nmimetype=> ${a?.mimetype}\nbitrate=> ${a?.bitrate}";
      info +=
          "\nheight=> ${a?.height}\nwidth=> ${a?.width}\nfileSize=> ${a?.filesize} Bytes\nduration=> ${a?.duration} milisec";
      info +=
          "\norientation=> ${a?.orientation}\ndate=> ${a?.date}\nframerate=> ${a?.framerate}";
      info += "\nlocation=> ${a?.location}";
    });

  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        floatingActionButtonLocation: FloatingActionButtonLocation.centerFloat,
        floatingActionButton: FloatingActionButton.extended(
            backgroundColor: Colors.white,
            label: const Text(
              "Get Info",
              style: TextStyle(color: Colors.black),
            ),
            icon: const Icon(
              Icons.video_call_outlined,
              color: Colors.purple,
            ),
            onPressed: () async => getVideoInfo(),
        ),
        appBar: AppBar(
          centerTitle: true,
          backgroundColor: Colors.purple,
          title: const Text('Video Info'),
        ),
        body: Padding(
          padding: const EdgeInsets.all(30.0),
          child: Center(
            child: Text(
              info,
              style: const TextStyle(fontSize: 21),
            ),
          ),
        ),
      ),
    );
  }
}


ios/Classes/SwiftVideoMetaInfoPlugin.swift
import AVFoundation
import Flutter
import MobileCoreServices
import UIKit

public class SwiftVideoMetaInfoPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "video_meta_info", binaryMessenger: registrar.messenger())
    let instance = SwiftVideoMetaInfoPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard call.method == "getVidInfo" else {
      result(FlutterMethodNotImplemented)
      return
    }

    guard let args = call.arguments else {
      return
    }

    if let myArgs = args as? [String: Any],
       let path = myArgs["path"] as? String
    {
      getVidInfo(path: path, result: result)
    }
  }

  private func getVidInfo(path: String, result: FlutterResult) {
    let url = URL(fileURLWithPath: path)
    let asset = AVURLAsset(url: url)
    let fileManager = FileManager.default
    let isFileExist = fileManager.fileExists(atPath: path)

    let creationDate = asset.creationDate?.dateValue

    let formatter = DateFormatter()
    formatter.timeZone = TimeZone.current
    formatter.locale = Locale.current
    formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
    let durationTime = round(CMTimeGetSeconds(asset.duration) * 1000)

    let tracks = asset.tracks(withMediaType: .video)
    let bitrate = round((tracks.first?.estimatedDataRate)!)
    let fps = tracks.first?.nominalFrameRate
    let size = tracks.first?.naturalSize
    let mimetype = mimeTypeForPath(path: path)
    let orientation = asset.orientation

    var fileSize: UInt64 = 0

    do {
      let attr = try FileManager.default.attributesOfItem(atPath: path)
      fileSize = attr[FileAttributeKey.size] as! UInt64
      let dict = attr as NSDictionary
      fileSize = dict.fileSize()
    } catch {
      print("Error: \(error)")
    }

    var jsonObj = [String: Any]()
    jsonObj["path"] = path
    jsonObj["mimetype"] = mimetype
    jsonObj["bitrate"] = bitrate
    jsonObj["author"] = ""
    if let date = creationDate {
      jsonObj["date"] = formatter.string(from: date)
    } else {
      jsonObj["date"] = ""
    }
    jsonObj["width"] = size?.width
    jsonObj["height"] = size?.height
    jsonObj["location"] = ""
    jsonObj["framerate"] = fps
    jsonObj["duration"] = durationTime
    jsonObj["filesize"] = fileSize
    jsonObj["orientation"] = orientation
    jsonObj["isfileexist"] = isFileExist

    do {
      let jsonData = try JSONSerialization.data(withJSONObject: jsonObj)
      let jsonStr = String(bytes: jsonData, encoding: .utf8)!
      result(jsonStr)
    } catch let e {
      result(e)
    }
  }

  private func mimeTypeForPath(path: String) -> String {
    let url = NSURL(fileURLWithPath: path)
    let pathExtension = url.pathExtension

    if let uti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, pathExtension! as NSString, nil)?.takeRetainedValue() {
      if let mimetype = UTTypeCopyPreferredTagWithClass(uti, kUTTagClassMIMEType)?.takeRetainedValue() {
        return mimetype as String
      }
    }
    return "application/octet-stream"
  }
}

extension AVURLAsset {
  var orientation: String {
    guard let tf = tracks(withMediaType: AVMediaType.video).first?.preferredTransform else {
      return ""
    }

    if tf.a == 0, tf.b == 1.0, tf.d == 0 {
      return "90"
    } else if tf.a == 0, tf.b == -1.0, tf.d == 0 {
      return "270"
    } else if tf.a == 1.0, tf.b == 0, tf.c == 0 {
      return "0"
    } else if tf.a == -1.0, tf.b == 0, tf.c == 0 {
      return "180"
    }
    return ""
  }
}


android/src/main/java/com/dstk/video_meta_info/VideoMetaInfoPlugin.java
package com.dstk.video_meta_info;


import android.content.Context;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.os.Build;

import androidx.annotation.NonNull;

import org.json.JSONObject;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * VideoMetaInfoPlugin
 */
public class VideoMetaInfoPlugin implements FlutterPlugin, MethodCallHandler {

    private String chName = "video_meta_info";
    public static Context context;

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        final MethodChannel channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(),
                "video_meta_info");
        channel.setMethodCallHandler(new VideoMetaInfoPlugin());
        context = flutterPluginBinding.getApplicationContext();
    }

    public static void registerWith(Registrar registrar_) {
        final MethodChannel channel = new MethodChannel(registrar_.messenger(), "video_meta_info");
        channel.setMethodCallHandler(new VideoMetaInfoPlugin());
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("getVidInfo")) {
            String path = call.argument("path");
            result.success(getVidInfo(path));
        } else {
            result.notImplemented();
        }
    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    }

    String getVidInfo(String path) {
        File file = new File(path);
        boolean isFileExists=file.exists();
        String author,dateString,mimeType,bitRate,location,frameRateStr,widthStr,heightStr,durationStr,orientation;
        double filesize;
        if(isFileExists){
            MediaMetadataRetriever mediaRetriever = new MediaMetadataRetriever();
            mediaRetriever.setDataSource(context, Uri.fromFile(file));

            author = getData(MediaMetadataRetriever.METADATA_KEY_AUTHOR, mediaRetriever);
            dateString = getData(MediaMetadataRetriever.METADATA_KEY_DATE, mediaRetriever);
            try {
                SimpleDateFormat readFormat = new SimpleDateFormat("yyyyMMdd'T'HHmmss.SSS", Locale.getDefault());
                Date date = readFormat.parse(dateString);
                SimpleDateFormat outFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault());
                dateString = outFormat.format(date);
            } catch (Exception e) {
                e.printStackTrace();
            }

            mimeType = getData(MediaMetadataRetriever.METADATA_KEY_MIMETYPE, mediaRetriever);
            bitRate = getData(MediaMetadataRetriever.METADATA_KEY_BITRATE, mediaRetriever);
            location = getData(MediaMetadataRetriever.METADATA_KEY_LOCATION, mediaRetriever);
            frameRateStr = getData(MediaMetadataRetriever.METADATA_KEY_CAPTURE_FRAMERATE, mediaRetriever);
            durationStr = getData(MediaMetadataRetriever.METADATA_KEY_DURATION, mediaRetriever);
            widthStr = getData(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH, mediaRetriever);
            heightStr = getData(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT, mediaRetriever);
            filesize = file.length();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                orientation = getData(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION, mediaRetriever);
            } else {
                orientation = null;
            }

            mediaRetriever.release();
        }else{
            author="";
            dateString="";
            mimeType="";
            bitRate="";
            location="";
            frameRateStr="";
            widthStr="";
            heightStr="";
            durationStr="";
            orientation="";
            filesize=0;
        }

        JSONObject json = new JSONObject();
        try {
            json.put("path", path);
            json.put("mimetype", mimeType);
            json.put("bitrate", bitRate);
            json.put("author", author);
            json.put("date", dateString);
            json.put("width", widthStr);
            json.put("height", heightStr);
            json.put("location", location);
            json.put("framerate", frameRateStr);
            json.put("duration", durationStr);
            json.put("filesize", filesize);
            json.put("orientation", orientation);
            json.put("isfileexist",isFileExists);
        } catch (Exception e) {
            e.printStackTrace();
        }

        return json.toString();
    }

    String getData(int key, MediaMetadataRetriever mediaRetriever) {
        try {
            return mediaRetriever.extractMetadata(key);
        } catch (Exception e) {
            return null;
        }
    }

}


lib/video_meta_info.dart

import 'dart:async';
import 'dart:convert';
import 'package:flutter/services.dart';
import 'package:path/path.dart';
import 'dart:developer' as developer;

class VideoMetaInfo {
  static const MethodChannel _channel = MethodChannel('video_meta_info');

  static Future<String?> get platformVersion async {
    final String? version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  Future<VideoData?> getVideoInfo(String path) async {
    final jsonStr = await _channel.invokeMethod('getVidInfo', {"path": path});
    final jsonMap = json.decode(jsonStr);
    if (!jsonMap["isfileexist"]) {
      developer.log('Video file path is wrong or does not exists. Please recheck file path.',
          name: 'VideoMetaInfo');
    }
    return VideoData.fromJson(jsonMap);
  }
}

class VideoData {
  String? path;

  /// string
  String? title;

  /// string
  String? author;

  /// string
  String? mimetype;

  /// Int
  int? bitrate;

  /// string
  String? date;

  /// string
  String? location;

  /// double
  double? framerate;

  /// int
  int? width;

  /// int
  int? height;

  /// [Android] API level 17, (0,90,180,270)
  /// (0 - LandscapeRight)
  /// (90 - Portrait)
  /// (180 - LandscapeLeft)
  /// (270 - portraitUpsideDown)
  int? orientation;

  /// bytes
  int? filesize;

  /// millisecond
  double? duration;

  VideoData({
    required this.path,
    this.title,
    this.author,
    this.mimetype,
    this.bitrate,
    this.date,
    this.location,
    this.framerate,
    this.width,
    this.height,
    this.orientation,
    this.filesize,
    this.duration,
  });

  VideoData.fromJson(Map<String, dynamic> json) {
    path = (json['path']);
    title = json["isfileexist"] ? (basename(json['path']).split(".")[0]) : null;
    mimetype = (json["mimetype"] == null || json["mimetype"] == "")
        ? null
        : (json["mimetype"]);
    bitrate = int.tryParse('${json['bitrate']}');
    date = (json["date"] == null || json["date"] == '') ? null : json["date"];
    location = (json["location"] == null || json["location"] == '')
        ? null
        : json["location"];
    framerate = double.tryParse("${json["framerate"]}");
    author = (json['author'] == null || json['author'] == '')
        ? null
        : json['author'];
    width = int.tryParse('${json['width']}');
    height = int.tryParse('${json['height']}');
    orientation = int.tryParse(json["orientation"]);
    filesize = json['filesize'];
    duration = json["isfileexist"] ? double.tryParse('${json['duration']}'):null;
  }
}
